### ISSUE TITLE
The highlighting code should be optimised

### ISSUE THREAD

**Issue by nad** [agda/agda#5326]
[Elsewhere](https://github.com/agda/agda/issues/4954#issuecomment-821480767) I wrote the following:

> [T]he code that finds the overlapping part could be slow if it is used incorrectly:

> https://github.com/agda/agda/blob/c880f8614f0c86ea0856ab0aacd456a4442218f7/src/full/Agda/Interaction/Highlighting/Precise.hs#L451-L468

> However, performance should be OK if we always remove the first part. (If necessary one could presumably optimise the code by using a different data structure.)

The `splitC` function is called from `generateAndPrintSyntaxInfo`:

https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/Interaction/Highlighting/Generate.hs#L163-L164

I added code that counted the length of `prevTokens`, and this indicates that the worst-case time complexity is at least quadratic in the size of the input file. For instance, consider the following code:
```agda
_ : Set₁
_ = Set

_ : Set₁
_ = Set

_ : Set₁
_ = Set

_ : Set₁
_ = Set

_ : Set₁
_ = Set

_ : Set₁
_ = Set
```
For this code we get the following list of lengths: [0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 0]. For "real" code the numbers can exceed 1000.

I suggest that we define `CompressedFile` using a data structure that allows us to implement `splitC` efficiently. One option might be to use a map with [`Range`s](https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/Interaction/Highlighting/Range.hs#L29-L36) as keys, along with the function [`splitLookup`](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Map-Strict.html#v:splitLookup).

**Comment by nad** [comment]
Perhaps the new representation can also be used for [`File`](https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/Interaction/Highlighting/Precise.hs#L197-L202), in which case [`CompressedFile`](https://github.com/agda/agda/blob/aea2c53340f509c2c67157caeac360400ea6afbc/src/full/Agda/Interaction/Highlighting/Precise.hs#L345-L350) could be removed.

**Comment by andreasabel** [comment]
5 years ago Guilhem Moulin was working on this, but this wasn't merged.  Maybe Guilhem still has the patches.  They were on a separate repo.  
But if you are fixing this now, @nad, there may be no need to look for the old solution.

I had also looked around for better representations.  There are some libraries for interval trees on hackage, but I didn't find one back than that implemented what I wanted.

**Comment by nad** [comment]
I've implemented some different data structures and run some benchmarks (`make std-lib-test`). Some observations and comments:
* I made the changes and ran the benchmarks on top of (roughly) commit f86f7044b40fc4a73e04254b627610062a295acb plus the change in commit 92c6f646364183e9f95ca238919b43df066b9033.
* Times vary quite a lot from run to run.
* Quite a bit of the time spent on computing and serialising highlighting data is not attributed to highlighting when `-vprofile:7` is used. I turned off highlighting by editing the code, and limited benchmarking suggests that, before my changes, computation and serialisation of highlighting data accounts for about 4-11% of the time.
* Given the lack of reliability of the benchmarks I am not entirely certain that the new implementation is faster than the previous one for the standard library.
* The problem I mentioned in the OP does not seem to be very important for the standard library. However, I constructed a file with 8580 occurrences of the following piece of code:
  ```agda
  _ : Set₁
  _ = Set
  ```
  That file took about four times longer to type-check without my changes.
* I've left several data structures in the code. (In fact, they are all used.) One data structure is used for the first step, where highlighting is gathered, and another for the second step, where it is modified. Some of the operations used for the second step are only implemented for one data structure (`RangeMap`), but several data structures can be used for the second operation by changing the type synonym `HighlightingInfoBuilder`.
* I did not find a Haskell implementation of an interval map type with only non-overlapping ranges and the kinds of operations I was looking for, so I implemented my own (`RangeMap`) using `Data.Map.Strict`.
* `RangeMap` could presumably be optimised further, but recompiling Agda and running benchmarks takes a lot of time.
* It might make sense to move some code to one or more modules under `Agda.Utils`.

**Comment by andreasabel** [comment]
Excellent!

>     * It might make sense to move some code to one or more modules under `Agda.Utils`.

Yes, I think it is good to separate the data structure (`RangeMap`) from its application (map to `Aspects`).

Please open a PR so we can have a structured discussion with the conveniences offered by the PR functionality.

**Comment by nad** [comment]
Another comment: The append function (`<>`) is not associative for `Aspects`. I think this is the reason for some changes to the highlighting behaviour. @andreasabel, you made append non-associative (66a82e62df9455e158818189d6ed3c6ee98f8626). I'm not convinced that this is a good idea.

**Comment by nad** [comment]
> I think this is the reason for some changes to the highlighting behaviour.

No, this was due to a mistake I made.

**Comment by nad** [comment]
I opened a separate issue related to how append is not associative (#5347).

The code passes all the tests now. It took me a few iterations to get the code to work for old versions of GHC. I also changed the implementation. Now the conversion from `HighlightingInfoBuilder` to `HighlightingInfo` no longer goes via `PositionMap`. (The difference in performance seems to be small for the standard library.)

I assume that it is possible to create a more efficient implementation of `RangeMap` by not reusing the implementation of `Data.Map.Strict`, but instead having a dedicated tree type with direct support for ranges. For instance, even in the case of no overlap the current implementation of `insert` uses four tree operations (unless the range is empty), which I guess translates to four tree traversals. It should be possible to traverse the tree just a single time.
